### PR TITLE
apt/deb: write esm apt sources on trusty. handle upgrade path to new series

### PR DESCRIPTION
**Pull request by blackboxsw**
Package postinst hook will now perform two tasks:
   - call configure_default_apt_sources to write ESM unauthenticated apt source file
   - call migrate_apt_sources to upgrade existing enabled entitlements to a new series
Package postrm will;
  - clean any apt source files written by uaclient
    
Fixes: #213

### CODE REVIEW

File: uaclient/apt.py [hunk 1]
```diff
@@ -170,3 +174,54 @@ def remove_apt_list_files(repo_url, series):
...
+        repo_url = ent_cls.repo_url
+        if cfg.is_attached:
+            resource_cfg = cfg.entitlements.get(ent_cls.name, {})
+            directives = resource_cfg.get(
+                'entitlement', {}).get('directives', {})
+            repo_url = directives.get('aptURL')
+        enabled = bool(re.search(r'(?P<pin>(-)?\d+) %s' % repo_url, policy))
+        if not os.path.exists(repo_filename):
+            if ent_cls.name == 'esm':
+                if platform_info['series'] == 'trusty':
```
Comment (powersj) [inline uaclient/apt.py]: Rather than hard-coding trusty, is there something we can check to ensure that series is supported?

Comment (powersj) [reply uaclient/apt.py]: I take this back, this is used for first install, so we wouldn't know what was or wasn't supported. Ignore this :)

Comment (blackboxsw) [reply uaclient/apt.py]: right, the only type of system we'd be able to check supported/not is after upgrade from a previous ubuntu-advantage-tools v19 or later which had already been attached. Then we'd have entitlement access information for support levels.

File: uaclient/apt.py [hunk 2]
```diff
@@ -170,3 +174,54 @@ def remove_apt_list_files(repo_url, series):
...
+
+def reconfigure_apt_sources(cfg=None, platform_info=None):
+    """Reconfigure apt sources for the given series.
+
+    Automatically setup an unauthenticated apt source for esm
+    production if not already present.
+    For each configured and enabled entitlement, update the sources files to
+    the new series.
+
+    @param platform_info: dict of platform information for testing
```
[the remark this answers is not in the document]
Comment (blackboxsw) [reply uaclient/apt.py]: added docstr

File: uaclient/apt.py (hunk 1 above)
Comment (blackboxsw) [reply uaclient/apt.py]: I've added a cfg.is_attached check where we loop over 'active' entitlements and re-enable that entitlement if the new series is reported as supported by ua-contracts

File: uaclient/apt.py [hunk 3]
```diff
@@ -170,3 +174,54 @@ def remove_apt_list_files(repo_url, series):
     for path in find_apt_list_files(repo_url, series):
         if os.path.exists(path):
             os.unlink(path)
+
+
+def reconfigure_apt_sources(cfg=None, platform_info=None):
```
[the remark this answers is not in the document]
Comment (blackboxsw) [reply uaclient/apt.py]: good point, maybe this esm portion should should be a separate script or just in the postinst/postrm directly as nothing else should call it. I could see migrate_apt_sources needing a bit more testing and needing to be called on enable or on forced disable when if client ever starts dealing with contract/resource expiry.

File: uaclient/apt.py [hunk 4]
```diff
@@ -170,3 +174,54 @@ def remove_apt_list_files(repo_url, series):
...
+            repo_url = directives.get('aptURL')
+        enabled = bool(re.search(r'(?P<pin>(-)?\d+) %s' % repo_url, policy))
+        if not os.path.exists(repo_filename):
+            if ent_cls.name == 'esm':
+                if platform_info['series'] == 'trusty':
+                    logging.info(
+                        'Providing unauthenticated ESM apt source file: %s',
+                        repo_filename)
+                    add_auth_apt_repo(repo_filename, repo_url)
+            elif enabled:  # And not esm
```
[the remark this answers is not in the document]
Comment (blackboxsw) [reply uaclient/apt.py]: If the service is(was) enabled on trusty and we upgraded to xenial, then the wrong repo_file exists
/etc/apt/sources.list.d/ubuntu-esm-trusty.list, and our new series check will be looking for /etc/apt/sources.list.d/ubuntu-esm-xenial.list. So we need to remove the old file and create a xenial source file in this case. This also helps cleanup ua-tools v18 which also wrote the /etc/apt/sources.list.d/ubuntu-entitlement-trusty.list  file as well).

File: uaclient/apt.py [hunk 5]
```diff
@@ -170,3 +174,54 @@ def remove_apt_list_files(repo_url, series):
...
+                logging.info('Upgrading existing "%s" apt source files: %s',
+                             ent_cls.title, repo_filename)
+
+                add_auth_apt_repo(repo_filename, repo_url)
```
[the remark this answers is not in the document]
Comment (blackboxsw) [reply uaclient/apt.py]: fixed

[the remark this answers is not in the document]
Comment (blackboxsw) [reply uaclient/apt.py]: dropped this as we can use entitlement.enable() directly

**Pull request merged**